compiler/rustc_next_trait_solver/src/solve/trait_goals.rs

```diff
@@ -133,31 +133,45 @@ where
         then: impl FnOnce(&mut EvalCtxt<'_, D>) -> QueryResult<I>,
     ) -> Result<Candidate<I>, NoSolution> {
         if let Some(trait_clause) = assumption.as_trait_clause() {
-            if trait_clause.def_id() == goal.predicate.def_id()
-                && trait_clause.polarity() == goal.predicate.polarity
-            {
+            let goal_did = goal.predicate.def_id();
+            let trait_clause_did = trait_clause.def_id();
+            if trait_clause_did == goal_did && trait_clause.polarity() == goal.predicate.polarity {
                 if !DeepRejectCtxt::relate_rigid_rigid(ecx.cx()).args_may_unify(
                     goal.predicate.trait_ref.args,
                     trait_clause.skip_binder().trait_ref.args,
                 ) {
                     return Err(NoSolution);
                 }
 
-                ecx.probe_trait_candidate(source).enter(|ecx| {
+                return ecx.probe_trait_candidate(source).enter(|ecx| {
                     let assumption_trait_pred = ecx.instantiate_binder_with_infer(trait_clause);
                     ecx.eq(
                         goal.param_env,
                         goal.predicate.trait_ref,
                         assumption_trait_pred.trait_ref,
                     )?;
                     then(ecx)
-                })
-            } else {
-                Err(NoSolution)
+                });
+            }
+
+            // PERF(sized-hierarchy): Sizedness supertraits aren't elaborated to improve perf, so
+            // check for a `Sized` subtrait when looking for `MetaSized`. `PointeeSized` bounds
+            // are syntactic sugar for a lack of bounds so don't need this.
+            if ecx.cx().is_lang_item(goal_did, TraitSolverLangItem::MetaSized)
+                && ecx.cx().is_lang_item(trait_clause_did, TraitSolverLangItem::Sized)
+                && {
+                    let expected_self_ty = ecx.resolve_vars_if_possible(goal.predicate.self_ty());
+                    let found_self_ty =
+                        ecx.resolve_vars_if_possible(trait_clause.self_ty().skip_binder());
+                    goal.predicate.polarity == trait_clause.polarity()
+                        && expected_self_ty == found_self_ty
```

Comment:
structurally equatingn (`==`) types does not handle alias types. Create a `<sized_self_type>: MetaSized` assumption instead and then recur.

With your impl, a `<T as Id>::This: Sized` where-clause does not imply `T: MetaSized`. Please add this as a test.